File: bishi/4-24华子/b.py
class Entry:
    def __init__(self, c, id):
        self.x = 0  # Count of '1's
        self.y = 0  # Maximum consecutive '1's
        self.c = c  # Original string
        self.id = id  # Identifier
        cnt = 0
        for ch in c:
            if ch == '1':
                cnt += 1
                self.x += 1
            else:
                cnt = 0
            self.y = max(self.y, cnt)

    def __lt__(self, other):
        if self.x != other.x:
            return other.x - self.x  # Sort by x in descending order
        elif self.y != other.y:
            return other.y - self.y  # Sort by y in descending order
        elif self.c != other.c:
            return self.c > other.c
        else:
            return self.id - other.id  # Sort by id in ascending order


def solve():
    n, m = map(int, input().split())
    s = list(input().split())
    a = []
    for i in range(n):
        entry = Entry(s[i], i + 1)
        a.append(entry)

    a.sort()

    for i in range(n - 1):
        print(a[i].id, end=" ")
    print(a[n - 1].id)


if __name__ == "__main__":
    solve()

File: bishi/4-24华子/b.cpp
#include <bits/stdc++.h>
using namespace std;

void solve()
{
	int n, m;
	cin >> n >> m;
	vector<string> s(n);
	vector<tuple<int, int, string, int>> a;

	for (int i = 0; i < n; ++i)
	{
		cin >> s[i];
		int x = 0, y = 0, cnt = 0;
		for (int j = 0; j < m; ++j)
		{
			if (s[i][j] == '1')
			{
				cnt++;
				x++;
			}
			else
			{
				cnt = 0;
			}
			y = max(y, cnt);
		}
		a.emplace_back(-x, -y, s[i], i + 1);
	}

	sort(a.begin(), a.end(), [](const auto &t1, const auto &t2)
		 {
        if (get<0>(t1) != get<0>(t2)) return get<0>(t1) < get<0>(t2);
        if (get<1>(t1) != get<1>(t2)) return get<1>(t1) < get<1>(t2);
        if (get<2>(t1) != get<2>(t2)) return get<2>(t1) > get<2>(t2);
        return get<3>(t1) < get<3>(t2); });
	for (int i = 0; i < n - 1; ++i)
	{
		std::cout << get<3>(a[i]) << " ";
	}
	std::cout << get<3>(a[n - 1]);
}

int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);

	solve();

	return 0;
}
